x-pack/plugins/siem/server/lib/hosts/elasticsearch_adapter.ts

```diff
@@ -4,67 +4,93 @@
+import { hostFieldsMap } from '../ecs_fields';
 import { FrameworkAdapter, FrameworkRequest, RequestOptions } from '../framework';
 import { TermAggregation } from '../types';
 
-import { buildQuery, hostsFieldsMap } from './query.dsl';
-import { HostBucket, HostData, HostHit, HostsAdapter } from './types';
+import { buildHostDetailsQuery } from './query.detail_host.dsl';
+import { buildHostsQuery } from './query.hosts.dsl';
+import { buildLastFirstSeenHostQuery } from './query.last_first_seen_host.dsl';
+import {
+  HostBuckets,
+  HostDetailData,
+  HostDetailEsItem,
+  HostDetailsRequestOptions,
+  HostEsData,
+  HostLastFirstSeenRequestOptions,
+  HostsAdapter,
+  HostValue,
+} from './types';
 
 export class ElasticsearchHostsAdapter implements HostsAdapter {
   constructor(private readonly framework: FrameworkAdapter) {}
 
   public async getHosts(request: FrameworkRequest, options: RequestOptions): Promise<HostsData> {
-    const response = await this.framework.callWithRequest<HostData, TermAggregation>(
+    const response = await this.framework.callWithRequest<HostEsData, TermAggregation>(
       request,
       'search',
-      buildQuery(options)
+      buildHostsQuery(options)
     );
     const { cursor, limit } = options.pagination;
     const totalCount = getOr(0, 'aggregations.host_count.value', response);
-    const hits: HostHit[] = getOr([], 'aggregations.group_by_host.buckets', response).map(
-      (bucket: HostBucket) => ({
-        ...head(bucket.host.hits.hits),
-        cursor: bucket.key!.host_name,
-        firstSeen: bucket.firstSeen.value_as_string,
-      })
-    );
-    const hostsEdges = hits.map(hit => formatHostsData(options.fields, hit, hostsFieldsMap));
+    const buckets: HostDetailEsItem[] = getOr([], 'aggregations.host_data.buckets', response);
+    const hostsEdges = buckets.map(bucket => formatHostEdgesData(options.fields, bucket));
     const hasNextPage = hostsEdges.length === limit + 1;
     const beginning = cursor != null ? parseInt(cursor, 10) : 0;
     const edges = hostsEdges.splice(beginning, limit - beginning);
 
+    return { edges, totalCount, pageInfo: { hasNextPage, endCursor: { value: String(limit) } } };
+  }
+
+  public async getHostDetails(
+    request: FrameworkRequest,
+    options: HostDetailsRequestOptions
+  ): Promise<HostItem> {
+    const response = await this.framework.callWithRequest<HostDetailData, TermAggregation>(
+      request,
+      'search',
+      buildHostDetailsQuery(options)
+    );
+    const aggregations: HostDetailEsItem = get('aggregations', response) || {};
+    return { _id: options.hostName, ...formatHostItem(options.fields, aggregations) };
+  }
+
+  public async getHostLastFirstSeen(
+    request: FrameworkRequest,
+    options: HostLastFirstSeenRequestOptions
+  ): Promise<LastFirstSeen> {
+    const response = await this.framework.callWithRequest<HostDetailData, TermAggregation>(
```

Comment:
Is `HostDetailData` the correct type for the `getHostLastFirstSeen` response?

Reply:
Nope